File: leetcode/dfs bfs/1034. Coloring A Border/py/main.py
class Solution:
    # Time O(N * M)
    # Space O(N * M)
    def colorBorder(self, grid, r0, c0, color):
        n, m = len(grid), len(grid[0])
        visited = [[False] * m for _ in range(n)]
        coords = []

        def neighbors(i, j):
            for x, y in [[i + 1, j], [i - 1, j], [i, j - 1], [i, j + 1]]:
                if 0 <= x < n and 0 <= y < m:
                    yield x, y

        def isBorder(i, j):
            for x, y in [[i + 1, j], [i - 1, j], [i, j - 1], [i, j + 1]]:
                if 0 <= x < n and 0 <= y < m and grid[x][y] != grid[i][j]:
                    return True
            return False

        def dfs(x, y):

            if x == 0 or y == 0 or x == n - 1 or y == m - 1 or isBorder(x, y):
                coords.append([x, y])

            visited[x][y] = True

            for nextX, nextY in neighbors(x, y):
                if not visited[nextX][nextY] and grid[nextX][nextY] == grid[
                        r0][c0]:
                    dfs(nextX, nextY)

        dfs(r0, c0)

        for x, y in coords:
            grid[x][y] = color

        return grid

File: leetcode/dfs bfs/1034. Coloring A Border/js/main.js
// Time O(N*M)
// Space O(N*M)
const colorBorder = (grid, r0, c0, color) => {
  let n = grid.length;
  let m = grid[0].length;
  let coords = [];
  let targetColor = grid[r0][c0];
  let coloredCoords = [];

  let visited = Array(n)
    .fill(0)
    .map(() => Array(m).fill(false));

  let dirs = [
    [0, 1],
    [0, -1],
    [-1, 0],
    [1, 0],
  ];

  dfs(r0, c0);

  for (let [i, j] of coords) {
    let isValid = false;

    for (let dir of dirs) {
      let x = i + dir[0];
      let y = j + dir[1];

      if (x < 0 || y < 0 || x >= n || y >= m) {
        isValid = true;
        break;
      } else {
        if (grid[x][y] != targetColor) {
          isValid = true;
          break;
        }
      }
    }

    if (isValid) {
      coloredCoords.push([i, j]);
    }
  }

  // Can color a square if and only if
  // 1) Square has a neighbor that is outside the grid
  // 2) Square has a neighbor whitch has a different color
  for (let [i, j] of coloredCoords) {
    grid[i][j] = color;
  }

  return grid;

  function dfs(i, j) {
    coords.push([i, j]);
    visited[i][j] = true;

    for (let dir of dirs) {
      let x = dir[0] + i;
      let y = dir[1] + j;

      if (x >= 0 && y >= 0 && x < n && y < m && grid[x][y] == targetColor && !visited[x][y]) {
        dfs(x, y);
      }
    }
  }
};
